### ISSUE TITLE
thunder grad transform is tracing backward number scalar inputs as proxies.

### ISSUE THREAD

**Issue by jjsjann123** [Lightning-AI/lightning-thunder#845]
A simple repro as below:

```
import thunder

def foo(a):
    b = a * 2.0
    return b

import torch

dtype = torch.float32
a = torch.randn(1, 3, 8, 8, device="cuda").to(dtype=dtype)
a.requires_grad_()

jfoo = thunder.jit(foo)

out = jfoo(a)
torch.autograd.backward(out, torch.rand_like(out), retain_graph=True)

print("====================")
print("\tcompute last trace:\n", thunder.last_traces(jfoo)[0])
print("\tcompute last backward trace:\n", thunder.last_backward_traces(jfoo)[0])
```

With the repro above, I'm seeing

```
====================
        compute last trace:
 import thunder
import thunder.torch as ltorch
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a):
  # a: "cuda:0 f32[1, 3, 8, 8]"

  # /volume/thunder_grad2.py:7:             b = a * 2.0
  result = ltorch.mul(a, 2.0)  # result: "cuda:0 f32[1, 3, 8, 8]"
    # result = prims.mul(a, 2.0)  # result: "cuda:0 f32[1, 3, 8, 8]"
  return result
        compute last backward trace:
 # Constructed by Backward pass
import thunder
import thunder.torch as ltorch
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, C1, = saved_for_backward
  t1, = cotangents
  # C0 (empty sequence)
  f1, = C1
  t7 = ltorch.mul(f1, t1)  # t7: "cuda:0 f32[1, 3, 8, 8]"
    # t7 = prims.mul(f1, t1)  # t7: "cuda:0 f32[1, 3, 8, 8]"
  return (t7,)
```

Notice that how the forward is seeing `result = ltorch.mul(a, 2.0)`, while backward sees:
```
  f1, = C1
  t7 = ltorch.mul(f1, t1)  # t7: "cuda:0 f32[1, 3, 8, 8]"
```

This is part of the reason we had https://github.com/Lightning-AI/lightning-thunder/pull/451#issuecomment-2191672425.

**Comment by jjsjann123** [comment]
Mistakenly thought the old issue #541 was for recompute dropout. Turns out it's exactly the same issue here. Sorry about the noise. I'll close this one and move the conversation back to the old issue.
